File: go/fiber-postgres/internal/middleware/metrics.go
package middleware

import (
	"time"

	"github.com/gofiber/fiber/v2"
	"go.opentelemetry.io/otel/attribute"

	"go-fiber-postgres/internal/telemetry"
)

func Metrics() fiber.Handler {
	return func(c *fiber.Ctx) error {
		start := time.Now()

		err := c.Next()

		duration := float64(time.Since(start).Milliseconds())
		status := c.Response().StatusCode()
		method := c.Method()
		path := c.Route().Path

		attrs := []attribute.KeyValue{
			attribute.String("http.method", method),
			attribute.String("http.route", path),
			attribute.Int("http.status_code", status),
		}

		telemetry.HTTPRequestsTotal.Add(c.UserContext(), 1, telemetry.WithAttributes(attrs...))
		telemetry.HTTPRequestDuration.Record(c.UserContext(), duration, telemetry.WithAttributes(attrs...))

		return err
	}
}

File: rust/actix-postgres/src/middleware/metrics.rs
use actix_web::body::MessageBody;
use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
use opentelemetry::KeyValue;
use std::future::{Future, Ready, ready};
use std::pin::Pin;
use std::time::Instant;

use crate::telemetry::{HTTP_REQUESTS_TOTAL, HTTP_REQUEST_DURATION};

pub struct MetricsMiddleware;

impl<S, B> Transform<S, ServiceRequest> for MetricsMiddleware
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error> + 'static,
    B: MessageBody + 'static,
{
    type Response = ServiceResponse<B>;
    type Error = actix_web::Error;
    type Transform = MetricsMiddlewareService<S>;
    type InitError = ();
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ready(Ok(MetricsMiddlewareService { service }))
    }
}

pub struct MetricsMiddlewareService<S> {
    service: S,
}

impl<S, B> Service<ServiceRequest> for MetricsMiddlewareService<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error> + 'static,
    B: MessageBody + 'static,
{
    type Response = ServiceResponse<B>;
    type Error = actix_web::Error;
    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;

    fn poll_ready(
        &self,
        ctx: &mut core::task::Context<'_>,
    ) -> core::task::Poll<Result<(), Self::Error>> {
        self.service.poll_ready(ctx)
    }

    fn call(&self, req: ServiceRequest) -> Self::Future {
        let method = req.method().to_string();
        let path = req.match_pattern().unwrap_or_else(|| req.path().to_string());
        let start = Instant::now();

        let fut = self.service.call(req);

        Box::pin(async move {
            let res = fut.await?;
            let status = res.status().as_u16().to_string();
            let duration = start.elapsed().as_secs_f64() * 1000.0;

            let labels = [
                KeyValue::new("http.method", method),
                KeyValue::new("http.route", path),
                KeyValue::new("http.status_code", status),
            ];

            HTTP_REQUESTS_TOTAL.add(1, &labels);
            HTTP_REQUEST_DURATION.record(duration, &labels);

            Ok(res)
        })
    }
}
